### PR TITLE
refactor SB* file uploader into top-level HOC

### PR DESCRIPTION
**Pull request by benjiwheeler**
### Resolves

Big step towards resolving https://github.com/LLK/scratch-gui/issues/4762

### Proposed Changes

* move `sb-file-uploader` to a top-level HOC that wraps `gui`
* in situation where user has selected to upload project, but file content is not ready/available, dispatch `project-state` action `RETURN_TO_SHOWING`

### Reason for Changes

1. we made design decision to use custom confirm modal, to have greater confidence in and control of the user experience when the user is at risk of overwriting their work. But this is not doable yet because clicking on the modal is interpreted as a click outside the open file menu, which closes the file menu. This is a problem because...
2. previously, the `sb-file-uploader` would be unmounted any time the file menu closed; its existence/mounting should not depend on a menu. 
3. in testing, when i would clear the contents of the file to upload before it could be loaded into the vm, `sb-file-uploader-hoc` would correctly avoid loading the empty content; but it would be stuck in the `LOADING_VM_FILE_UPLOAD` state. Submitting `RETURN_TO_SHOWING` in this case makes sure we will stay in the right state in that case. (We shouldn't actually be missing the contents of the file to upload in practice, but just in case!)

### Test Coverage

Added a test and moved over tests to new HOC, but ideally we would test even more.

### CODE REVIEW

File: src/reducers/project-state.js [hunk 1]
```diff
@@ -435,10 +435,21 @@ const onLoadedProject = (loadingState, canSave, success) => {
...
+    } else {
+        switch (loadingState) {
+        case LoadingState.LOADING_VM_WITH_ID:
+        case LoadingState.LOADING_VM_FILE_UPLOAD:
+            return {
+                type: RETURN_TO_SHOWING
+            };
+        case LoadingState.LOADING_VM_NEW_DEFAULT:
+            return {
+                type: START_ERROR
```
Comment (benjiwheeler) [inline src/reducers/project-state.js]: If loading fails (success === false), recover by going back to just showing project as-is. But if we were loading the default project, a failure in loading probably indicates a serious error.

File: test/unit/util/project-saver-hoc.test.jsx [hunk 2]
```diff
@@ -56,7 +56,7 @@ describe('projectSaverHOC', () => {
         expect(mockedUpdateProject).toHaveBeenCalled();
     });
 
-    test('if canSave is alreatdy true and we show a project with an id, project will NOT be saved', () => {
+    test('if canSave is already true and we show a project with an id, project will NOT be saved', () => {
```
Comment (benjiwheeler) [inline test/unit/util/project-saver-hoc.test.jsx]: fixed a typo

File: src/lib/sb-file-uploader-hoc.jsx [hunk 3]
```diff
@@ -0,0 +1,241 @@
+import bindAll from 'lodash.bindall';
```
Comment (benjiwheeler) [inline src/lib/sb-file-uploader-hoc.jsx]: I tried to keep as much as possible in this file the same as before. There are very few differences, aside from what was necessary to refactor it into a top-level HOC wrapping GUI.

**Comment by benjiwheeler** [comment]
Ready for review; just blocking until after this week's deploy.

**Comment by benjiwheeler** [comment]
@paulkaplan and I talked about this. Some takeaways:

Concerns:
* Z is currently working to start loading project before gui fully renders. Adding a top level HOC, with an additionally rendered input component, can slow this down.
* need to assess/measure/understand how much this increases the overall performance burden on gui.

Possible alternative approaches:
* in `containers/menu.jsx`, `mouseup` events cause the menu to be closed. This is the original problem leading to this PR's refactor. What if instead, we just checked if the mouse target was a modal, and didn't close the menu then?

Next step:
* pick this back up when we have more time to discuss. Possibly with @rschamp .

Labeling as blocked, per need for discussion.

File: src/reducers/project-state.js [hunk 4]
```diff
@@ -435,10 +435,21 @@ const onLoadedProject = (loadingState, canSave, success) => {
         default:
             return;
         }
+    } else {
```
Comment (rschamp) [inline src/reducers/project-state.js]: Can this be extracted into a different PR?

File: src/lib/sb-file-uploader-hoc.jsx [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+                /* eslint-enable no-unused-vars */
+                ...componentProps
+            } = this.props;
+            return (
+                <React.Fragment>
+                    <WrappedComponent
+                        onStartSelectingFileUpload={this.handleStartSelectingFileUpload}
+                        {...componentProps}
+                    />
+                    {fileInput}
```
Comment (rschamp) [inline src/lib/sb-file-uploader-hoc.jsx]: Let's try to create this element dynamically in the handler, rather than mounting it here, to avoid issues with components unmounting.

**Comment by benjiwheeler** [comment]
Discussed with @rschamp and @paulkaplan. Current plan: 

1. The menu click issue that prompted this should be fixed in and of itself
2 I should pull out my project state tests and state fix, and introduce those separately
3. The custom modal for project upload can then be introduced
4. Redo this PR as only a refactor, which doesn't change behavior at all
    * change placement of the file picker to be a createelement that’s not even on the dom, per Ray's comment: "Let's try to create this element dynamically in the handler, rather than mounting it here, to avoid issues with components unmounting."

File: src/reducers/project-state.js (hunk 4 above)
Comment (benjiwheeler) [reply src/reducers/project-state.js]: Did so -- see https://github.com/LLK/scratch-gui/pull/5291

**Comment by benjiwheeler** [comment]
Updated plan:

1. Redo this PR as only a refactor, which doesn't change behavior at all.
    * change placement of the file picker to be a createelement that’s not even on the dom, per Ray's comment: "Let's try to create this element dynamically in the handler, rather than mounting it here, to avoid issues with components unmounting."
2. Re-introduce custom modal; discuss with participants on https://github.com/LLK/scratch-gui/issues/5386

File: src/lib/sb-file-uploader-hoc.jsx (hunk 3 above)
Comment (benjiwheeler) [reply src/lib/sb-file-uploader-hoc.jsx]: > I tried to keep as much as possible in this file the same as before. There are very few differences, aside from what was necessary to refactor it into a top-level HOC wrapping GUI.

Now there are more differences -- as part of making the <input> component dynamic, I reordered functions and added comments.

**Comment by benjiwheeler** [comment]
This PR is ready for review, but tests aren't passing; it's difficult to diagnose because the same tests pass locally.

**Comment by benjiwheeler** [comment]
OK @rschamp, take a look at the dynamic way sb-file-uploader-hoc is now creating the <input> element.

File: src/lib/sb-file-uploader-hoc.jsx [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+                    .then(() => {
+                        this.props.onLoadingFinished(this.props.loadingState, true);
+                        if (filename) {
+                            const uploadedProjectTitle = this.getProjectTitleFromFilename(filename);
+                            this.props.onUpdateProjectTitle(uploadedProjectTitle);
+                        }
+                    })
+                    .catch(error => {
+                        log.warn(error);
+                        this.props.intl.formatMessage(messages.loadError); // eslint-disable-line no-alert
```
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/lib/sb-file-uploader-hoc.jsx]: Doing it

File: src/lib/sb-file-uploader-hoc.jsx [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+        const loadingState = state.scratchGui.projectState.loadingState;
+        const user = state.session && state.session.session && state.session.session.user;
+        return {
+            isLoadingUpload: getIsLoadingUpload(loadingState),
+            isShowingWithoutId: getIsShowingWithoutId(loadingState),
+            loadingState: loadingState,
+            projectChanged: state.scratchGui.projectChanged,
+            userOwnsProject: ownProps.authorUsername && user &&
+                (ownProps.authorUsername === user.username),
+            vm: state.scratchGui.vm // NOTE: double check this belongs here
```
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/lib/sb-file-uploader-hoc.jsx]: Doing it
